packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -725,14 +725,24 @@ class RenderParagraph extends RenderBox with ContainerRenderObjectMixin<RenderBo
   @override
   bool hitTestChildren(BoxHitTestResult result, { required Offset position }) {
     final TextPosition textPosition = _textPainter.getPositionForOffset(position);
-    final Object? span = _textPainter.text!.getSpanForPosition(textPosition);
-    if (span is HitTestTarget) {
-      result.add(HitTestEntry(span));
-      return true;
+    if (_positionIsWithinText(position: position, textPosition: textPosition)) {
+      final InlineSpan? span = _textPainter.text!.getSpanForPosition(textPosition);
+      if (span is HitTestTarget) {
+        result.add(HitTestEntry(span! as HitTestTarget));
+        return true;
+      }
     }
     return hitTestInlineChildren(result, position);
   }
 
+  bool _positionIsWithinText({ required Offset position, required TextPosition textPosition }){
+    final List<ui.TextBox> boxes = _textPainter.getBoxesForRange(
+      math.max(0, textPosition.offset - 1),
+      math.min(_textPainter.plainText.length, textPosition.offset + 1),
+    );
+    return boxes.where((ui.TextBox box) => box.toRect().contains(position)).isNotEmpty;
```

Comment:
`boxes.any` so it stops when it finds a box.

Reply:
I suspect this is going to be slow. There's a new Skia API that will allow us to do this correctly and efficiently. Unfortunately it's not exposed in text painter yet. If this fix is not needed immediately would it be ok to wait until the new APIs are available?

Reply:
This is not needed immediately, I can close this pr and wait for the new APIs.

Reply:
@LongCatIsLooong What is the timeline on the Skia API?  I have a need for this in Linkify (https://github.com/flutter/flutter/pull/125927).  I filed https://github.com/flutter/flutter/issues/131435 and verified that this PR did fix the issue for me locally.

Even though I'm excited for this to fix my Linkify problem, it's not really super urgent.  Linkify is not yet ready to merge, and merging it without this fix wouldn't even be terrible.  I just want to make sure it's not going to be months or something for the Skia change.

Reply:
The skia API is already available it just needs to be exposed (but the HTML renderer needs a separate implementation). I'll start working on that once the rounding thing is removed.

Reply:
Sounds good, thanks!